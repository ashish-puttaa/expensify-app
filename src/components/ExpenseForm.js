import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// const date = new Date();
// console.log('date', date);

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
   state = {
      description: '',
      note: '',
      amount: '',
      createdAt: moment(),
      calendarFocused: false,
      error: ''
   };

   componentDidMount() {
      if (this.props.expense) {
         const expense = this.props.expense;

         this.setState(() => ({
            description: expense.description,
            note: expense.note,
            amount: (expense.amount / 100).toString(),
            createdAt: moment(expense.createdAt)
         }));
      }
   }

   onAmountChange = e => {
      const amount = e.target.value;
      if (!amount || amount.match(/^\d+(\.\d{0,2})?$/))
         this.setState(() => ({ amount }));
   };

   onDateChange = createdAt => {
      if (createdAt) this.setState(() => ({ createdAt }));
   };

   onDescriptionChange = e => {
      const description = e.target.value;
      this.setState(() => ({ description }));
   };

   onFocusChange = ({ focused }) => {
      this.setState(() => ({ calendarFocused: focused }));
   };

   onNoteChange = e => {
      const note = e.target.value;
      this.setState(() => ({ note }));
   };

   onSubmit = e => {
      e.preventDefault();

      if (!this.state.description && !this.state.amount) {
         this.setState(() => ({
            error: '*Please provide Description and Amount'
         }));
      } else if (!this.state.description) {
         this.setState(() => ({
            error: '*Please provide Description'
         }));
      } else if (!this.state.amount) {
         this.setState(() => ({
            error: '*Please provide Amount'
         }));
      } else {
         this.setState(() => ({ error: '' }));

         this.props.onSubmit({
            description: this.state.description,
            amount: parseFloat(this.state.amount, 10) * 100,
            createdAt: this.state.createdAt.valueOf(),
            note: this.state.note
         });
      }
   };

   render() {
      return (
         <div>
            <form onSubmit={this.onSubmit}>
               <input
                  type="text"
                  placeholder="Description"
                  autoFocus
                  value={this.state.description}
                  onChange={this.onDescriptionChange}
               />
               <input
                  type="text"
                  placeholder="Amount"
                  value={this.state.amount}
                  onChange={this.onAmountChange}
               />
               <SingleDatePicker
                  date={this.state.createdAt}
                  onDateChange={this.onDateChange}
                  focused={this.state.calendarFocused}
                  onFocusChange={this.onFocusChange}
                  numberOfMonths={1}
                  isOutsideRange={day => false}
               />
               <textarea
                  placeholder="Add a note for your expense (optional)"
                  value={this.state.note}
                  onChange={this.onNoteChange}
               ></textarea>
               <button>
                  {this.props.expense ? 'Edit Expense' : 'Add Expense'}
               </button>
            </form>
            {this.state.error && <p>{this.state.error}</p>}
         </div>
      );
   }
}
