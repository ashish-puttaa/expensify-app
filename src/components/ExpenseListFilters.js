import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from './../actions/filters';

/* 
   Exported ONLY to use it as an UNCONNECTED version in Snapshot Testing
   unconnected -> not connecting to react-redux using connect()
*/
export class ExpenseListFilters extends React.Component {
   state = {
      focusedInput: null
   };

   onDatesChange = ({ startDate, endDate }) => {
      this.props.setStartDate(startDate);
      this.props.setEndDate(endDate);
   };

   onFocusChange = focusedInput => {
      this.setState(() => ({ focusedInput }));
   };

   onTextChange = e => {
      this.props.setTextFilter(e.target.value);
   };

   onSortChange = e => {
      if (e.target.value === 'date') {
         this.props.sortByDate();
      } else if (e.target.value === 'amount') {
         this.props.sortByAmount();
      }
   };

   render() {
      return (
         <div>
            <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
            <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
               <option value="date">Date</option>
               <option value="amount">Amount</option>
            </select>
            <DateRangePicker
               startDate={this.props.filters.startDate}
               endDate={this.props.filters.endDate}
               onDatesChange={this.onDatesChange}
               focusedInput={this.state.focusedInput}
               onFocusChange={this.onFocusChange}
               hideKeyboardShortcutsPanel={true}
               numberOfMonths={1}
               isOutsideRange={() => false}
               showClearDates={true}
            />
         </div>
      );
   }
}

const mapStateToProps = state => {
   return {
      filters: state.filters
   };
};

const mapDispatchToProps = dispatch => ({
   setStartDate: startDate => dispatch(setStartDate(startDate)),
   setEndDate: endDate => dispatch(setEndDate(endDate)),
   setTextFilter: text => dispatch(setTextFilter(text)),
   sortByDate: () => dispatch(sortByDate()),
   sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
