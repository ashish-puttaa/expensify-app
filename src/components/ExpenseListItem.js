import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const centsToDollars = amount => {
   return `$${(amount / 100).toFixed(2)}`;
};

const millisecondsToDate = timeInMs => {
   return moment(timeInMs).format('MMM Do, YYYY');
};

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
   <div>
      <Link to={`/edit/${id}`}>
         <h3>{description}</h3>
      </Link>

      <p>
         {centsToDollars(amount)} - {millisecondsToDate(createdAt)}
      </p>
   </div>
);

export default ExpenseListItem;
