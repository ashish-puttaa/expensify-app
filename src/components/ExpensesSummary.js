import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import getExpenesesTotal from './../selectors/expenses-total';
import selectExpenses from './../selectors/expenses';
import numeral from 'numeral';

const centsToDollars = amount => {
   // return `$${(amount / 100).toFixed(2)}`;
   return numeral(amount / 100).format('$0,0.00');
};

export const ExpensesSummary = props => {
   const expenseWord = props.expensesCount === 1 ? 'expense' : 'expenses';
   const expensesTotalInDollars = centsToDollars(props.expensesTotal);

   return (
      <div className="page-header">
         <div className="content-container">
            <h1 className="page-header__title">
               Viewing <span>{props.expensesCount}</span> {expenseWord} totalling <span>{expensesTotalInDollars}</span>
            </h1>
            <div className="page-header__actions">
               <Link className="button" to="/create">
                  Add Expense
               </Link>
            </div>
         </div>
      </div>
   );
};

const mapStateToProps = state => {
   const filteredExpenses = selectExpenses(state.expenses, state.filters);
   return {
      expensesCount: filteredExpenses.length,
      expensesTotal: getExpenesesTotal(filteredExpenses)
   };
};

export default connect(mapStateToProps)(ExpensesSummary);
