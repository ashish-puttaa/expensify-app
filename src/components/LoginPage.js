import React from 'react';
import { startLogin } from './../actions/auth';
import { connect } from 'react-redux';

export const LoginPage = props => (
   <div>
      <h1>Login Page</h1>
      <button onClick={props.startLogin}>Login</button>
   </div>
);

const mapDispatchToProps = dispatch => ({
   startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
