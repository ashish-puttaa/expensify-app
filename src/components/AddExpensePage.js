import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from './../actions/expenses';

/* 
   Exported ONLY to use it as an UNCONNECTED version in Snapshot Testing
   unconnected -> not connecting to react-redux using connect()
*/
export class AddExpensePage extends React.Component {
   onSubmit = expense => {
      /* 
         this.props.dispatch() can be called by creating a spy during testing, 
         but addExpense cannot be called that way as it is being imported.
         To fix this, mapDispatchToProps.
         mapDispatchToProps removes the dispatch function from the props, and instead adds the new function provided.
         The new function can be defined to use the imported method, which makes easier to create spies.

         this.props.dispatch(addExpense(expense));
      */

      this.props.startAddExpense(expense);
      this.props.history.push('/');
   };

   render() {
      return (
         <div>
            <div className="page-header">
               <div className="content-container">
                  <h2 className="page-header__title">Add Expense</h2>
               </div>
            </div>

            <div className="content-container">
               <ExpenseForm onSubmit={this.onSubmit} />
            </div>
         </div>
      );
   }
}

// To make it easier to use spies during testing
const mapDispatchToProps = dispatch => ({
   startAddExpense: expense => dispatch(startAddExpense(expense))
});

/* export default connect()(AddExpensePage); */

export default connect(undefined, mapDispatchToProps)(AddExpensePage);

/*














*/
// Before converting to Class based Component to avoid inline functions:
// NOTE :
//    You can just create functions inside the stateless components,
//    but the disadvantage is that the function has to be re-created every time the component is rendered
//
//
// const AddExpensePage = props => (
//    <div>
//       <h2>Add Expense</h2>
//       <ExpenseForm
//          onSubmit={expense => {
//             console.log(props);
//
//             /* props.dispatch() can be called by creating a spy during testing,
//                but addExpense cannot be called that way as it is being imported.
//                To fix this, mapDispatchToProps.
//                mapDispatchToProps removes the dispatch function from the props, and instead adds the new function provided.
//                The new function can be defined to use the imported method, which makes easier to create spies.
//
//                props.dispatch(addExpense(expense));
//             */
//
//             props.addExpense(expense);
//             props.history.push('/');
//          }}
//       />
//    </div>
// );
