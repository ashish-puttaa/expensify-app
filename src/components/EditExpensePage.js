import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from './../actions/expenses';

/* 
   Exported ONLY to use it as an UNCONNECTED version in Snapshot Testing
   unconnected -> not connecting to react-redux using connect()
*/
export class EditExpensePage extends React.Component {
   onSubmit = expense => {
      /* this.props.dispatch(editExpense(this.props.expense.id, expense)); */
      this.props.startEditExpense(this.props.expense.id, expense);
      this.props.history.push('/');
   };

   onRemove = () => {
      /* 
         this.props.dispatch() can be called by creating a spy during testing, 
         but removeExpense cannot be called that way as it is being imported.
         To fix this, mapDispatchToProps.
         mapDispatchToProps removes the dispatch function from the props, and instead adds the new function provided.
         The new function can be defined to use the imported method, which makes easier to create spies.

         this.props.dispatch(removeExpense({ id: this.props.expense.id }));
      */

      this.props.startRemoveExpense({ id: this.props.expense.id });
      this.props.history.push('/');
   };

   render() {
      return (
         <div>
            <div className="page-header">
               <div className="content-container">
                  <h2 className="page-header__title">Edit Expense</h2>
               </div>
            </div>
            <div className="content-container">
               <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
               <button className="button button--secondary" onClick={this.onRemove}>
                  Remove Expense
               </button>
            </div>
         </div>
      );
   }
}

const mapStateToProps = (state, props) => {
   return {
      expense: state.expenses.find(expense => {
         return expense.id === props.match.params.id;
      })
   };
};

// To make it easier to use spies during testing
const mapDispatchToProps = dispatch => ({
   startRemoveExpense: idObj => dispatch(startRemoveExpense(idObj)),
   startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense))
});

/* export default connect(mapStateToProps)(EditExpensePage); */

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

/*















*/

// Before converting to Class based Component to avoid inline functions:
// NOTE :
//    You can just create functions inside the stateless components,
//    but the disadvantage is that the function has to be re-created every time the component is rendered
//
//
// const EditExpensePage = props => {
//    return (
//       <div>
//          <ExpenseForm
//             expense={props.expense}
//             onSubmit={expense => {
//                /* props.dispatch(editExpense(props.expense.id, expense)); */
//                props.editExpense(props.expense.id, expense);
//                props.history.push('/');
//             }}
//          />
//          <button
//             onClick={e => {
//                /*
//                   props.dispatch() can be called by creating a spy during testing,
//                   but removeExpense cannot be called that way as it is being imported.
//                   To fix this, mapDispatchToProps.
//                   mapDispatchToProps removes the dispatch function from the props, and instead adds the new function provided.
//                   The new function can be defined to use the imported method, which makes easier to create spies.
//
//                   props.dispatch(removeExpense({ id: props.expense.id }));
//                */
//
//                props.removeExpense({ id: props.expense.id });
//                props.history.push('/');
//             }}
//          >
//             Remove
//          </button>
//       </div>
//    );
// };
