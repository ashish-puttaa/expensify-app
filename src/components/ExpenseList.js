import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from './../selectors/expenses';

// Exported ONLY to use it as an UNCONNECTED version in Snapshot Testing
// unconnected -> not connecting to react-redux using connect()
export const ExpenseList = props => (
   <div>
      <h2>Expense List</h2>
      {props.expenses.length === 0 ? (
         <p>No expenses</p>
      ) : (
         props.expenses.map(expense => (
            <ExpenseListItem key={expense.id} {...expense} />
         ))
      )}
   </div>
);

const mapStateToProps = state => {
   return {
      expenses: selectExpenses(state.expenses, state.filters)
   };
};

export default connect(mapStateToProps)(ExpenseList);

/* Without using selectors : 

   const ConnectedExpenseList = connect(state => {
      return {
         expenses: state.expenses
      };
   })(ExpenseList);

   export default ConnectedExpenseList;
*/
