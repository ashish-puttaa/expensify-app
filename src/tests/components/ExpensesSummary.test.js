import React from 'react';
import { ExpensesSummary } from './../../components/ExpensesSummary';
import { shallow } from 'enzyme';

test('should render ExpensesSummary with multiple expenses', () => {
   const wrapper = shallow(
      <ExpensesSummary expensesCount={21} expensesTotal={782364} />
   );
   expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with one expense', () => {
   const wrapper = shallow(
      <ExpensesSummary expensesCount={1} expensesTotal={1024} />
   );
   expect(wrapper).toMatchSnapshot();
});
