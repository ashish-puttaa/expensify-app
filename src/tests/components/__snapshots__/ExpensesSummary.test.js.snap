// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h4>
    Viewing 21 expenses totalling $7,823.64
  </h4>
</div>
`;

exports[`should render ExpensesSummary with one expense 1`] = `
<div>
  <h4>
    Viewing 1 expense totalling $10.24
  </h4>
</div>
`;
